### DISCUSSION TITLE
mockoon-cli does not return response body

### DISCUSSION THREAD

**Discussion by nleeuskadi** [mockoon/mockoon#1107] [Help] [answered]
Hi,

I an using mockoon-cli docker image with this command
`docker run -d --mount type=bind,source=/my-path/my-mockoon-file-env.json,target=/data,readonly -p 3333:3333 mockoon/cli:latest -d data -p 3333` and I ma facing an issue with the response body which are empty

OS = Ubuntu 22.04.2 LTS

The mockoon-cli starts successfully and the routes end-points I configured in the provided mockoon file are reachable (I used the desktop mockoon applicaiton to design my routes).
I can see this in the logs
```
$ docker logs ab125d0ee550
{"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"f312b390-74dc-4bea-85e8-6ac2ecf81b8b","level":"info","message":"Server started on port 4444","timestamp":"2023-08-16T09:02:47.392Z"}
{"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"f312b390-74dc-4bea-85e8-6ac2ecf81b8b","level":"info","message":"Transaction recorded","requestMethod":"GET","requestPath":"/workorder('7006873')","requestProxied":false,"responseStatus":200,"timestamp":"2023-08-16T09:03:00.578Z"}
{"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"f312b390-74dc-4bea-85e8-6ac2ecf81b8b","level":"info","message":"Transaction recorded","requestMethod":"HEAD","requestPath":"/workorder('7006873')","requestProxied":false,"responseStatus":200,"timestamp":"2023-08-16T09:03:07.127Z"}

```

The problem is that the response body are empty and I don't understand why?
When typing this command `curl http://localhost:4444/workorder\(\'7006873\'\)` a response body shall be returned, but nothing is returned.

Also, if I only request headers (`curl -I http://localhost:4444/workorder\(\'7006873\'\)`), All the expected headers of the routes are not returned. That is very strange.

For information, everything works fine if i use the Desktop mockoon application: both response body and headers are well returned.

Did you ever face this issue?

Thank you very much for your help

Cheers

**Comment by 255kb** [comment]
Hi,

This sounds strange as the CLI is "just" a wrapper around Mockoon's core which is shared between the two apps. What works in the desktop app should work in the CLI too. 
I suspect that something is wrong with the parentheses in the path as they are [considered as a regular expression by Express](https://expressjs.com/en/guide/routing.html).
Could you show how you declared the route path? And maybe [share the full environment](https://mockoon.com/docs/latest/mockoon-data-files/environment-clipboard-copy/#copy-an-environment-or-a-route-json-to-the-clipboard)? I will try to reproduce the issue.

Thanks

**Reply by nleeuskadi** [reply]
Hi Guillaume,

thank you very much for your answer :)
 Your idea with the parentheses sounds interesting.
I know this type of API is not common, but the system we have to mock expose this type of REST signature :(
I attached my environment json file.

Here below the results of different tests I executed:
 - the localhost:3000 is the route exposed by mockoon desktop application and it works fine
 - the localhost:3333 is the same route exposed by mockoon-cli and it does not work as expected (response code is OK, but not the headers and body response neither). Mockoon-cli launched with this command: `docker run -d --mount type=bind,source=/fullpath/my-mock-api.json,target=/data,readonly -p 3333:3333 mockoon/cli:latest -d data -p 3333`
 
 ```
 $ curl http://localhost:3000/workorder\(\'7006873\'\)
{
  "response": "welcome!!"
}

$ curl http://localhost:3333/workorder\(\'7006873\'\)
$ curl -I http://localhost:3000/workorder\(\'7006873\'\)
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET,POST,PUT,PATCH,DELETE,HEAD,OPTIONS
Access-Control-Allow-Headers: Content-Type, Origin, Accept, Authorization, Content-Length, X-Requested-With
Content-Security-Policy: default-src *; style-src 'self' 'unsafe-inline'; font-src 'self' data:; script-src 'self' 'unsafe-inline' 'unsafe-eval' *
Content-Length: 29
Date: Wed, 16 Aug 2023 11:39:41 GMT
Connection: keep-alive
Keep-Alive: timeout=5

$ curl -I http://localhost:3333/workorder\(\'7006873\'\)
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 0
Date: Wed, 16 Aug 2023 11:39:50 GMT
Connection: keep-alive
Keep-Alive: timeout=5
 ```
 
 The logs of mockoon-cli:
 ```
 {"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"b2177939-dc18-45fb-93d7-72e3e432bc03","level":"info","message":"Server started on port 3333","timestamp":"2023-08-16T11:38:47.505Z"}
{"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"b2177939-dc18-45fb-93d7-72e3e432bc03","level":"info","message":"Transaction recorded","requestMethod":"GET","requestPath":"/workorder('7006873')","requestProxied":false,"responseStatus":200,"timestamp":"2023-08-16T11:39:16.825Z"}
{"app":"mockoon-server","environmentName":"my mock api","environmentUUID":"b2177939-dc18-45fb-93d7-72e3e432bc03","level":"info","message":"Transaction recorded","requestMethod":"HEAD","requestPath":"/workorder('7006873')","requestProxied":false,"responseStatus":200,"timestamp":"2023-08-16T11:39:50.029Z"}
[my-mock-api.json.gz](https://github.com/mockoon/mockoon/files/12359643/my-mock-api.json.gz)

**Reply by 255kb** [reply]
It seems that you shared an OpenAPI export file. I suggest that you read [this article](https://mockoon.com/docs/latest/openapi/openapi-specification-compatibility/) as importing from/exporting to OpenAPI is not a perfect 1:1 equivalent to what Mockoon is offering. 
I suspect that you ran the exported OpenAPI file in the CLI. This is supported but is more limited as you lose some information during the export (because there is no OAS equivalent or it's not yet supported). 
The CLI should consume your [Mockoon data file](https://mockoon.com/docs/latest/mockoon-data-files/data-storage-location/#locating-the-files) directly without the export step.

**Comment by nleeuskadi** [comment]
[my-mock-api.json.gz](https://github.com/mockoon/mockoon/files/12359652/my-mock-api.json.gz)

**Reply by nleeuskadi** [reply]
Hi Guillaume,

To complete my previous answer, I tried with another environment which only expose a basic /test route.
And I also reproduce my pb. The first idea related to the parenthesis may not be the reason of the issue.
Cheers.

**Reply by 255kb** [reply]
Could you please provide the exact steps/commands/files you did/use so I can try to reproduce? 

If you gave this JSON file `[my-mock-api.json.gz](https://github.com/mockoon/mockoon/files/12359652/my-mock-api.json.gz)` to the CLI, my previous explanation still stands.

**Comment by nleeuskadi** [comment, marked as answer]
Hi Guillaume,

I sent to you a gzip file because I cannot attach a json file directly in the github HMI.
Did you receive the gzipo file and did you managed to gunzip it?

What i did is :
1. launch the mockoon-cli docker image with the provided json file:
```
docker run -d --mount type=bind,source=/my-path/my-mock-api.json,target=/data,readonly -p 3333:3333 mockoon/cli:latest -d data -p 3333
```

2. execute the following curl command:
```
curl http://localhost:3333/workorder\(\'7006873\'\)
```
The expected response should be
```
{
  "response": "welcome!!"
}
```
But it returns nothing.

3. I requested all the headers for the same route
```
curl -I http://localhost:3333/workorder\(\'7006873\'\)
```
The expected response should display all the headers:
```
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET,POST,PUT,PATCH,DELETE,HEAD,OPTIONS
Access-Control-Allow-Headers: Content-Type, Origin, Accept, Authorization, Content-Length, X-Requested-With
Content-Security-Policy: default-src *; style-src 'self' 'unsafe-inline'; font-src 'self' data:; script-src 'self' 'unsafe-inline' 'unsafe-eval' *
Content-Length: 29
Date: Wed, 16 Aug 2023 11:39:41 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```
But it only returns some default headers
```
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 0
Date: Wed, 16 Aug 2023 12:16:37 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

Hope it can help you.
If you need more details or if something not clear please tell me.
Cheers

**Reply by nleeuskadi** [reply]
again, it is to be noticed that If I execute the same tests  with the same route (same environment json file) but exposed by Mockoon Desktop application, it works like a charm

**Reply by 255kb** [reply]
> Did you receive the gzipo file and did you managed to gunzip it?

Yes, and I will give you the same answer. 
You sent me an exported OAS file. You probably used this feature:

![image](https://github.com/mockoon/mockoon/assets/7489814/fb226c24-26de-4d45-ba43-bf0ffe4b5bc8)

This is **not** what you should use with the CLI. 

The CLI should directly load the Mockoon file that you will find by clicking here:

![image](https://github.com/mockoon/mockoon/assets/7489814/13a36fb2-6087-4f5f-a235-1eb30ed48f7d)

If you use the export feature you will **loose** information as Mockoon features and OAS are not a 1:1 equivalent. This is explained in the documentation [here](https://mockoon.com/docs/latest/openapi/openapi-specification-compatibility/).

So, in your case yes, you mays not get the same result in the desktop app and in the CLI as you are not using the same setup.

**Reply by nleeuskadi** [reply]
Hi Guillaume,

thank you very much for your explanation.
It solves my problem :)
I did read the mentioned document but obviously I did not understand the differences between the 2 features.
Sorry for that !

And thank you again for your patience and your help!
Cheers

**Reply by 255kb** [reply]
No problem, it may not be obvious... 
Hopefully, the more we improve the OAS import/export system, the less differences we will have. Still, some of Mockoon's features have no equivalent in OAS and vice versa. So, I guess it will never be a perfect way to share something or run the mock with the CLI.
